api/src/main/java/org/asynchttpclient/request/body/generator/ReactiveStreamBodyGenerator.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2010-2012 Sonatype, Inc. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package org.asynchttpclient.request.body.generator;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.asynchttpclient.request.body.Body;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReactiveStreamBodyGenerator implements BodyGenerator {
+    private static final ByteBuffer EMPTY = ByteBuffer.wrap("".getBytes());
+
+    private final Publisher<ByteBuffer> publisher;
+
+    public ReactiveStreamBodyGenerator(Publisher<ByteBuffer> publisher) {
+        this.publisher = publisher;
+    }
+
+    public Publisher<ByteBuffer> getPublisher() {
+        return this.publisher;
+    }
+
+    @Override
+    public Body createBody() {
+        return new StreamedBody(publisher);
+    }
+
+    private class StreamedBody implements Body {
+        private final QueueBackedSubscriber subscriber;
+
+        public StreamedBody(Publisher<ByteBuffer> publisher) {
+            this.subscriber = new QueueBackedSubscriber();
+            publisher.subscribe(subscriber);
+        }
+
+        @Override
+        public void close() throws IOException {
+            // noop
+        }
+
+        @Override
+        public long getContentLength() {
+            return -1L;
+        }
+
+        @Override
+        public State read(ByteBuffer buffer) throws IOException {
+            boolean isCompleted = subscriber.readChunk(buffer);
+            return (isCompleted) ? State.Stop : State.Continue;
+        }
+    }
+
+    private static class QueueBackedSubscriber implements Subscriber<ByteBuffer> {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(QueueBackedSubscriber.class);
+
+        private final BlockingQueue<ByteBuffer> queue;
```

Comment:
The alternative is queuing up elements in the subscriber, but the risk with that is overflowing the queue (dropping elements is even worse) when the consumer is much slower than the producer. Also, note this implementation is only used when the provider doesn't support reactive streams.